### PR TITLE
ci: docker build publish

### PR DESCRIPTION
**Pull request by tnightengale**
**PR Checklist**

- [x] A description of the changes is added to the description of this PR.
- [x] If there is a related issue, make sure it is linked to this PR.
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added or modified a feature, documentation in `docs` is updated

**Description of changes**

This PR adds a GitHub Actions workflow to build the `Dockerfile` and `uc-cli.dockerfile` images on changes to those files, building on the work scoped in #319.

An example of the built and pushed image produced by this workflow can be found here: https://hub.docker.com/r/tnightengale/unitycatalog.

**Perquisites to Merging** 
- [ ] Decide on multi-arch support 
- [ ] Refine run triggers to align with release process
- [ ] Add docs

### CODE REVIEW

File: .github/workflows/docker.yml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      with:
+        username: ${{ vars.DOCKERHUB_USERNAME }}
+        password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+    - name: Build and push Docker image
+      uses: docker/build-push-action@v6
+      with:
+        context: .
+        file: ${{ matrix.dockerfile }}
+        push: true
```
Comment (tnightengale) [inline .github/workflows/docker.yml:51]:
```suggestion
        push: false
```

To be removed; included to demonstrate pushing to my personal DockerHub. This PR should only add builds and any other relevant tests for the images. Pushing and tagging should be decided later.

File: .github/workflows/docker.yml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        platforms: linux/amd64 # TODO: add support for other platforms, after resolving: https://github.com/unitycatalog/unitycatalog/issues/480
```
Comment (tnightengale) [inline .github/workflows/docker.yml:52]: Note that our Dockerfiles seem to be using base images that are incompatible with supporting arm64, per  #480.

File: .github/workflows/docker.yml [hunk 3]
```diff
@@ -0,0 +1,57 @@
+name: Build and Push Docker Images
+
+on:
+  push:
+    # TODO: run on tags rather than branches
```
Comment (tnightengale) [inline .github/workflows/docker.yml:5]: Would love input here on when to trigger the builds.

File: .github/workflows/docker.yml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      uses: docker/build-push-action@v6
+      with:
+        context: .
+        file: ${{ matrix.dockerfile }}
+        push: true
+        platforms: linux/amd64 # TODO: add support for other platforms, after resolving: https://github.com/unitycatalog/unitycatalog/issues/480
+        tags: ${{ vars.DOCKERHUB_USERNAME }}/${{ matrix.image }}:latest
+
+        # NOTE: See here for an explanation of registry caching: https://tinyurl.com/4b8mfexr
+        cache-from: type=registry,ref=${{ vars.DOCKERHUB_USERNAME }}/${{ matrix.image }}:buildcache
```
Comment (tnightengale) [inline .github/workflows/docker.yml:56]: Caching is implemented at a registry level currently (ie. DockerHub). Here's a helpful article on why that may be a good approach: https://www.blacksmith.sh/blog/cache-is-king-a-guide-for-docker-layer-caching-in-github-actions

**Comment by jalotra** [comment]
@tnightengale this github action has failed previous runs, can you please check once ?

**Comment by tnightengale** [comment]
@jalotra Not sure what you mean exactly :) This PR is closed, and it's current form can be found at #551

**Pull request closed without merging**